File: vignettes/precompute.R
#!/usr/bin/env Rscript
#
# Precompute ("bake") the model-fitting tutorial vignettes.
#
# The tutorials fit real Stan models. To keep
# the documentation site fast to build and free of a live CmdStan dependency,
# each tutorial is authored as `<name>.Rmd.orig` (the source, with live fits)
# and this script knits it to `<name>.Rmd` with all model output and figures
# baked in. The committed `<name>.Rmd` and its figures are what pkgdown renders.
#
# Run this WHENEVER the modelling code (R/, inst/stan/) or a `.Rmd.orig` changes,
# then commit the regenerated `.Rmd` files and figures:
#
#     Rscript vignettes/precompute.R            # bake all tutorials
#     Rscript vignettes/precompute.R flu        # bake a subset
#
# Requires a working CmdStan installation (cmdstanr::install_cmdstan()).

args <- commandArgs(trailingOnly = TRUE)
force <- "--force" %in% args
args <- setdiff(args, "--force")
all_vigs <- c("flu", "multiple-obs", "multilevel-multi-obs",
              "flaxman", "b117")
vigs <- if (length(args)) intersect(args, all_vigs) else all_vigs
if (!length(vigs)) stop("No known tutorial requested. Choose from: ",
                        paste(all_vigs, collapse = ", "))

# locate the package root (this script lives in <root>/vignettes)
root <- tryCatch(
  rprojroot::find_root(rprojroot::is_r_package),
  error = function(e) normalizePath(".")
)
if (!file.exists(file.path(root, "DESCRIPTION"))) {
  stop("Run this from the package root: Rscript vignettes/precompute.R")
}

# load the in-development package so the fits use the current source
suppressWarnings(suppressMessages(pkgload::load_all(root, quiet = TRUE)))

vig_dir <- file.path(root, "vignettes")
owd <- setwd(vig_dir)
on.exit(setwd(owd), add = TRUE)

library(knitr)
# keep bookdown figure labels/anchors so \@ref(fig:...) cross-references survive
opts_knit$set(bookdown.internal.label = TRUE)

# A tutorial is stale when its baked .Rmd is older than its source, or older
# than anything in R/ or inst/stan/ that could change the numbers. Everything
# else is skipped, so re-running this after an unrelated edit costs nothing.
# `--force` bakes regardless.
model_mtime <- suppressWarnings(max(file.mtime(c(
  list.files(file.path(root, "R"), full.names = TRUE, pattern = "\\.[Rr]$"),
  list.files(file.path(root, "inst", "stan"), full.names = TRUE,
             recursive = TRUE),
  # the shipped data objects carry the delay kernels, so a change there moves
  # every fitted number even though no code changed
  list.files(file.path(root, "data"), full.names = TRUE)
)), -Inf))

is_stale <- function(src, out) {
  if (force || !file.exists(out)) return(TRUE)
  file.mtime(out) < max(file.mtime(src), model_mtime)
}

# knitr invalidates a cached chunk when ITS OWN code changes, not when the data
# underneath it does. A change to data/ or R/ is therefore replayed from cache
# and the vignette reports stale numbers under new prose -- which is exactly
# what happened when the inf2death kernel was corrected. --force drops the whole
# cache, since chunk labels need not begin with the vignette's name (europe-covid
# labels its chunks `multilevel-*`) and any per-vignette pattern would miss them.
if (force && dir.exists("cache")) {
  n <- length(list.files("cache"))
  unlink("cache", recursive = TRUE)
  message("== Dropped the knitr cache (", n, " files) ==")
}

skipped <- character()
for (v in vigs) {
  src <- paste0(v, ".Rmd.orig")
  out <- paste0(v, ".Rmd")
  if (!file.exists(src)) { warning("missing ", src); next }
  if (!is_stale(src, out)) {
    message("== Skipping ", v, " (up to date; --force to rebake) ==")
    skipped <- c(skipped, v)
    next
  }
  message("== Baking ", v, " ==")
  # namespace each vignette's figures so they do not collide
  opts_chunk$set(fig.path = file.path("figure", paste0(v, "-")))
  knit(input = src, output = out, quiet = TRUE)
  message("   wrote ", out)
}

baked <- setdiff(vigs, skipped)
if (length(baked)) {
  # Record the input fingerprint so `make docs-check` can tell, in about a
  # second and without fitting anything, whether the published tutorials still
  # match the code. Only meaningful once EVERY tutorial has been baked, so skip
  # it for a partial run rather than record a stamp that overstates freshness.
  if (setequal(vigs, all_vigs)) {
    stamp <- file.path(root, "tools", "docs-stamp.sh")
    if (file.exists(stamp)) system2(stamp, c("write", "r"))
  } else {
    message("NOTE: partial bake (", paste(baked, collapse = ", "),
            ") -- docs stamp NOT updated. Run without arguments to refresh it.")
  }
  message("Done. Commit the regenerated .Rmd files and vignettes/figure/*.png")
} else {
  message("Nothing to do -- all requested tutorials are up to date.")
}

File: python/scripts/precompute.py
#!/usr/bin/env python
"""Precompute ("bake") the tutorial notebooks into the documentation site.

The Python counterpart of ``vignettes/precompute.R``, and it exists for the same
reason: the tutorials fit real Stan/PyMC models, and a documentation build should
not. Each tutorial is authored as ``notebooks/<name>.py`` (a jupytext percent
notebook, so it is also a runnable script); this script executes it once and
writes the rendered markdown and figures into ``docs/tutorials/``, which is what
mkdocs publishes.

    uv run --group dev python scripts/precompute.py                 # all
    uv run --group dev python scripts/precompute.py multilevel-multi-obs

Run it whenever the modelling code or a notebook changes, then commit the
regenerated markdown and images. Nothing is fitted at ``mkdocs build`` time.
"""

from __future__ import annotations

import argparse
import os
import shutil
import subprocess
import sys
import tempfile
import time
from contextlib import contextmanager
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
NOTEBOOKS = ROOT / "notebooks"
OUT = ROOT / "docs" / "tutorials"

# Order matters only for the printed summary.
ALL_TUTORIALS = ["flu", "europe-covid", "multiple-obs",
                 "partial-pooling", "multilevel-multi-obs", "flaxman", "b117"]

TITLES = {
    "europe-covid": "Assessing the effects of interventions",
    "flaxman": "Reproducing Flaxman et al. (2020)",
    "b117": "Transmissibility of a new variant",
    "partial-pooling": "Partial pooling",
    "multilevel-multi-obs": "Multilevel models with several observation series",
}


KERNEL = "epidemia-precompute"


@contextmanager
def project_kernel():
    """A throwaway kernelspec pointing at THIS interpreter.

    Without it, jupytext/nbconvert resolve the notebook's declared ``python3``
    kernel to whatever is registered globally -- which is not the project
    virtualenv, so the notebook fails on ``import plotnine``. Installing into a
    temporary prefix and pointing ``JUPYTER_PATH`` at it keeps the user's real
    kernel list untouched; the directory goes away on exit.
    """
    tmp = tempfile.mkdtemp(prefix="epidemia-kernel-")
    try:
        subprocess.run(
            [sys.executable, "-m", "ipykernel", "install",
             "--prefix", tmp, "--name", KERNEL],
            check=True, capture_output=True,
        )
        env = dict(os.environ)
        share = Path(tmp) / "share" / "jupyter"
        env["JUPYTER_PATH"] = (
            f"{share}{os.pathsep}{env['JUPYTER_PATH']}"
            if env.get("JUPYTER_PATH") else str(share)
        )
        yield env
    finally:
        shutil.rmtree(tmp, ignore_errors=True)


def bake(name: str, timeout: int, env: dict) -> bool:
    """Execute one notebook and write markdown + figures into docs/tutorials."""
    src = NOTEBOOKS / f"{name}.py"
    if not src.exists():
        print(f"  !! {src} does not exist", file=sys.stderr)
        return False

    OUT.mkdir(parents=True, exist_ok=True)
    ipynb = NOTEBOOKS / f"{name}.ipynb"

    print(f"== baking {name} ==", flush=True)
    started = time.time()
    try:
        # jupytext converts the percent script to a notebook and runs it; the
        # kernel is this project's interpreter, so the package under development
        # is what gets exercised.
        subprocess.run(
            ["jupytext", "--to", "ipynb", "--execute",
             "--set-kernel", KERNEL, "-o", str(ipynb), str(src)],
            cwd=ROOT, check=True, timeout=timeout, env=env,
        )
        # nbconvert writes <name>.md plus a <name>_files/ directory of figures.
        subprocess.run(
            ["jupyter", "nbconvert", "--to", "markdown",
             "--output-dir", str(OUT), str(ipynb)],
            cwd=ROOT, check=True, timeout=600, env=env,
        )
    except subprocess.CalledProcessError as exc:
        print(f"  !! {name} failed (exit {exc.returncode})", file=sys.stderr)
        return False
    except subprocess.TimeoutExpired:
        print(f"  !! {name} timed out after {timeout}s", file=sys.stderr)
        return False
    finally:
        # The .ipynb is an intermediate: the .py is the source of truth and the
        # .md is the artefact. Leaving it behind invites editing the wrong file.
        ipynb.unlink(missing_ok=True)

    _prepend_title(OUT / f"{name}.md", TITLES.get(name, name))
    print(f"   wrote docs/tutorials/{name}.md  ({time.time() - started:.0f}s)")
    return True


def _prepend_title(path: Path, title: str) -> None:
    """Give mkdocs a top-level heading even if the notebook starts with prose."""
    if not path.exists():
        return
    text = path.read_text()
    if text.lstrip().startswith("# "):
        return
    path.write_text(f"# {title}\n\n{text}")


def _newest(*paths) -> float:
    """Newest mtime among the given files and directory trees."""
    best = 0.0
    for p in paths:
        p = Path(p)
        if p.is_dir():
            best = max([best] + [f.stat().st_mtime for f in p.rglob("*")
                                 if f.is_file()])
        elif p.is_file():
            best = max(best, p.stat().st_mtime)
    return best


def is_stale(name: str) -> bool:
    """Does this tutorial need rebaking?

    Stale when the published page is missing, or older than its notebook source
    or than anything in the package that could change the numbers. Skipping the
    rest means re-running this after an unrelated edit costs nothing -- these
    notebooks fit real models and take minutes to hours.
    """
    out = OUT / f"{name}.md"
    if not out.exists():
        return True
    return out.stat().st_mtime < _newest(NOTEBOOKS / f"{name}.py",
                                         ROOT / "src" / "epidemia")


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("tutorials", nargs="*", default=None,
                    help="subset to bake; default is all of them")
    ap.add_argument("--timeout", type=int, default=3600,
                    help="per-notebook execution timeout in seconds")
    ap.add_argument("--clean", action="store_true",
                    help="remove docs/tutorials before baking")
    ap.add_argument("--force", action="store_true",
                    help="rebake even when the output is already up to date")
    args = ap.parse_args()

    for tool in ("jupytext", "jupyter"):
        if shutil.which(tool) is None:
            raise SystemExit(
                f"{tool} not found. Install the dev group: uv sync --group dev"
            )

    wanted = args.tutorials or ALL_TUTORIALS
    unknown = set(wanted) - set(ALL_TUTORIALS)
    if unknown:
        raise SystemExit(
            f"unknown tutorial(s) {sorted(unknown)}; "
            f"choose from {ALL_TUTORIALS}"
        )

    if args.clean and OUT.exists():
        shutil.rmtree(OUT)

    stale = [n for n in wanted if args.clean or args.force or is_stale(n)]
    for name in wanted:
        if name not in stale:
            print(f"== Skipping {name} (up to date; --force to rebake) ==")
    if not stale:
        print("\nNothing to do -- all requested tutorials are up to date.")
        return

    with project_kernel() as env:
        ok = [bake(name, args.timeout, env) for name in stale]
    failed = [n for n, good in zip(stale, ok) if not good]
    if failed:
        raise SystemExit(f"failed: {', '.join(failed)}")

    # Record the input fingerprint so `make docs-check` can answer "are the
    # published tutorials stale?" in about a second, without fitting anything.
    # Only when the FULL set was requested -- a stamp written after a partial
    # bake would claim more freshness than it has.
    if set(wanted) == set(ALL_TUTORIALS):
        stamp = ROOT.parent / "tools" / "docs-stamp.sh"
        if stamp.exists():
            subprocess.run([str(stamp), "write", "python"], check=False)
    else:
        print(f"NOTE: partial bake ({', '.join(wanted)}) -- docs stamp NOT "
              "updated. Run with no arguments to refresh it.")
    print("\nDone. Commit docs/tutorials/ and its *_files/ directories.")


if __name__ == "__main__":
    main()
